## Test dataset

We have provided a small 1KGP dataset for test runs. They are located in `test-dataset` directory. It includes the following:
- [1KGP.EUR.CEU.BGI.cram.ftp.test.list](1KGP.EUR.CEU.BGI.cram.ftp.test.list) : A list of FTP locations for 16 European CRAM files
- [1KGP.EUR.CEU.BGI.bam.example.list](1KGP.EUR.CEU.BGI.bam.example.list) : A list of example input file-paths for the 16 BAM files
- [example.loci.hg38.bed](example.loci.hg38.bed) : A list of four duplicated loci and their genomic positions
- [exons.hg38.noalt.bed](exons.hg38.noalt.bed) : A BED file containing genomic positions for ~185,000 exons

The `prior` directory contains prior copy number distributions for the four duplicated loci. 

The `out-depth` directory contains [all.counts.tsv](out-depth/all/all.counts.tsv), which is a matrix of background read counts mapped to the ~185,000 exons for the 16 example European samples. This process has been precomputed for the users to save time.

Please see [test_pipeline.sh](test_pipeline.sh) in this directory for the full test pipeline. You can run the full pipeline by running:
```
bash test_pipeline.sh
```

or by following step-by-step, as shown below:

## Pipeline description

#### Download hg38 reference genome (takes ~5 minutes)
```
wget -c -O hg38.fa ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/GRCh38_reference_genome/GRCh38_full_analysis_set_plus_decoy_hla.fa
wget -c -O hg38.fa.fai ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/GRCh38_reference_genome/GRCh38_full_analysis_set_plus_decoy_hla.fa.fai
```

#### Download Parascopy's hg38 homology table
```
wget -c https://dl.dropboxusercontent.com/s/okzeedb6gze6zzs/homology_table_hg38.tar
tar xf homology_table_hg38.tar
```

#### Download 16 example European samples (takes ~20 minutes using 8 cores)
```
mkdir -p cram-files
cat 1KGP.EUR.CEU.BGI.cram.ftp.list | xargs -n 1 -P 8 wget -P cram-files
```

#### Convert CRAM to BAM files (takes ~30 minutes using 8 cores)

```
ls cram-files/*.CEU.exome.cram | awk -F '.' '{print $1}' | awk -F '/' '{print $2}' | xargs -n 1 -P 8 bash cram_to_bam.sh
```

#### Make a list of input BAM paths
```
bash create_bam_paths.sh > 1KGP.EUR.CEU.BGI.bam.list
```

#### Run Edgecopy modules
```
# (1) depth module (~2 minutes using 8 cores)
edgecopy depth \
 --input 1KGP.EUR.CEU.BGI.bam.list \
 --output out-depth \
 --exon-list exons.hg38.noalt.bed \
 --reference hg38.fa \
 --hom-table homology_table/hg38.bed.gz \
 -@ 8

# (2) agCN module (~2 minutes using 8 cores)
edgecopy agcn \
 --input 1KGP.EUR.CEU.BGI.bam.list \
 --depth out-depth \
 --output out-agcn \
 --loci-list example.loci.hg38.bed \
 --exon-list exons.hg38.noalt.bed \
 --reference hg38.fa \
 --hom-table homology_table/hg38.bed.gz \
 --priors priors \
 -@ 8

# (3) psCN module (~3 minutes using 8 cores)
edgecopy pscn \
--input 1KGP.EUR.CEU.BGI.bam.list \
--output out-pscn \
--loci-list example.loci.hg38.bed \
--reference hg38.fa \
--hom-table homology_table/hg38.bed.gz \
--agcn-dir out-agcn \
-@ 8
```

## Expected output

The [expected-output](./expected-output/) directory contains sample output directories and files you can use to ensure the software is running correctly.

More specifically, `expected-output/out-depth`, `expected-output/out-agcn`, and `expected-output/out-pscn` are sample directories generated by running `edgecopy depth`, `edgecopy agcn`, and `edgecopy pscn`, respectively.